README.md
# Containerized Open Source Probo.CI Server
The construction of an open source ProboCI server within a Docker container.

The goal of this project is to have a fully functional Open Source Probo.CI container with a companion Drupal 8 based portal container for administering and looking at builds on the Open Source ProboCI Server.

### Complete Documentation Wiki
For complete information on this project including expanded documentation, please visit the Wiki page for this project on GitHub located at:  

https://github.com/ElusiveMind/probo-docker/wiki  

You can view additional documentation on the Probo Open Source initiative at:  

https://docs.probo.ci/open-source/  

### Configuration Environment Variable

**ASSET_RECEIVER_TOKEN**  
_Default: null_  
The token used for uploading assets to the receiver. Needs to be part of the request to push files to the asset receiver. If not specified, no token will be required which is a security issue.  

**ASSET_RECEIVER_URL**  
_Default: http://example.com:3070_  
The URL (with port) of your asset receiver process. This needs to be the URL with domain name and not an IP address as this will be looked up from within Docker containers. Probo uses this URL to fetch your assets from the asset receiver as part of the build process.  

**AWS_ACCESS_KEY_ID**  
_Default: null_  
When AwsS3Storage is configured as the asset receiver storage method, this is the access key id for the account used to store files.  

**AWS_ENDPOINT**  
_Default: null_  
If you have a custom end point that is used for something like a MinIO server, you can specify it here.  

**AWS_SECRET_ACCESS_KEY**  
_Default: null_  
When AwsS3Storage is configured as the asset receiver storage method, this is the secret access key for the account used to store files.  

**AWS_BUCKET**  
_Default: null_  
When AwsS3Storage is configured as the asset receiver storage method, this is the name of the bucket where the files are stored.  

**BB_ACCESS_TOKEN**  
_Default: null_  
The Bitbucket access token for this application. This must be obtained through the process laid out in the Wiki documentation using the Bitbucket configuration generation feature of the Drupal 8 module.  

**BB_CLIENT_KEY**  
_Default: null_  
The Bitbucket client key configured in your Bitbucket application.  

**BB_CLIENT_SECRET**  
_Default: null_  
The Bitbucket client secret configured in your Bitbucket application.  

**BB_REFRESH_TOKEN**  
_Default: null_  
The Bitbucket refresh token for this application. This must be obtained through the process laid out in the Wiki documentation using the Bitbucket configuration generation feature of the Drupal 8 module.  

**BB_WEBHOOK_URL**  
_Default: /bitbucket-webhook_  
The endpoint of the Bitbucket webhook. This is used in the path configured in your webhook on Bitbucket.  

**BYPASS_TIMEOUT**  
_Default: 0_  
A flag as to whether or not probo docker containers should timeout after a period of time. Setting this to 1 prevents container timeouts.  

**CM_INSTANCE_NAME**  
_Default: OSProboCI_  
The name of your ProboCI instance. This will be used as part of the context build string viewable on your code repository status screen. This should be different for each probo server you run and should be custom set for your instance.  

**CREATE_LOOM_TASK_LOG**  
_Default: 0_  
Determines whether individual logs for Probo loom tasks are created. Good for debugging, bad for production.  

**ENCRYPTION_CIPHER**  
_Default: aes-256-cbc_  
Determines the cipher used to encrypt the assets. See https://www.openssl.org/docs/manmaster/apps/ciphers.html for options.  

**ENCRYPTION_PASSWORD**  
_Default: password_  
A salt for ENCRYPTION_CIPHER.  

**FILE_STORAGE_PLUGIN**  
_Default: LocalFiles_  
This is for the asset receiver. This defines the plugin that will be used for file storage, Currently there are two options, LocalFiles and AwsS3Storage. AwsS3Storage requires the keys and a bucket for storing files on Amazon S3. Be sure NOT to put your docker-compose.yml file or other configuration file in any kind of public source repository with this data in it. It's bad. Very, very bad. Trust me.  

**GITHUB_API_TOKEN**  
_Default: personal token here_  
The GitHub API token configured in your GitHub account.  

**GITHUB_WEBHOOK_PATH**
_Default: /github-webhook_  
The endpoint of the GitHub webhook. This is used in the path configured in your webhook on GitHub.  

**GITHUB_WEBHOOK_SECRET**  
_Default: null_  
This value should be modified to a secure string as well. This is a random token containing a string value that you select and will need to be used when you configure your webhook in GitHub.  

**GITLAB_CLIENT_KEY**  
_Default: null_  
The GitLab client key configured in your GitLab application.  

**GITLAB_CLIENT_SECRET**  
_Default: null_  
The GitLab client secret configured in your GitLab application.  

**GITLAB_WEBHOOK_URL**  
_Default: /gitlab-webhook_  
The endpoint of the GitLab webhook. This is used in the path configured in your webhook on GitLab.  

**LOOM_SERVER_TOKEN**  
_Default: null_  
A token used for accessing the loom server. The loom server is currently used to provide event information to a web service. The web service calls the loom server using this token for all requests. If not configured, no token will be required to make requests. It is a best practice to configure a unique server token.

**LOOM_EVENT_API_URL**  
_Default: null_  

**PROBO_BUILD_URL**  
_Default: http://{{buildId}}.example.com:3050/_  
URL template string for viewing each build. {{buildId}} expands to the id of the build. Must include the port number if not using a standard port number (port 80).  

**PROXY_CACHE_ENABLED**  
_Default: true_  
Enable a caching mechanism for the proxy server. This will cache items locally for the proxy and result in better performance.  

**PROXY_CACHE_MAX**  
_Default: 500_  
Max number of proxy lookup responses to cache.  

**PROXY_CACHE_MAX_AGE**  
_Default: 5m_  
Max age of proxy lookup responses in cache, in any unit (units default to ms).  

**PROXY_HOSTNAME_IP**  
_Default: localhost_  
The host name where docker lives. Note that this should be the intranet IP address of where your root docker installation lives. Without this, proxying will fail.  

**PROBO_LOGGING**  
_Default: 0_  
Boolean value (either 1 or 0). Setting a value of 1 will cause each enabled process to log output. These logs are typically used for debugging builds and are not recommended for production server configurations.  

**PROXY_PORT**  
_Default: 3050_  
Port that the proxy server is running on.  

**PROXY_SERVER_TIMEOUT**  
_Default: 10m_  
Server timeout, in any unit.  

**REAPER_DRY_RUN**  
_Default: true_  

**REAPER_OUTPUT_FORMAT**  
_Default: text_  
The output format used for --status command. Options are:
- json: JSON output - one line per JSON object
- text: (default) - human readable hearchical output

**REAPER_BRANCH_BUILD_LIMIT**  
_Default: 1_  
Limit the number of builds per branch of code.  

**RECIPHERED_OUTPUT_DIR**  
_Default: null_  

**REDIRECT_URL**  
_Default: null_  
The URL to send people too when a build cannot be found or is in the process of being built. Also used for reaped builds. Sends query string at the end.  

**SERVICE_ENDPOINT_URL**  
_Default: http://www.example.com/probo-api/service-endpoint.json_  
An optional configuration for a URL to receive notification of service events. A JSON object will be sent to this URL with each build object containing information about the current build in process including tasks. This is designed specifically for use with the Drupal 8 module for the Probo Open Source Server available at: https://github.com/ElusiveMind/probo-drupal  

**UPLOADS_PAUSED**  
_Default: null_  

**USE_BITBUCKET**  
_Default: 0_  
A boolean value to communicate whether or not the Bitbucket handler should be configured and started as part of the server.  

**USE_GITLAB**  
_Default: 0_  
A boolean value to communicate whether or not the GitLab handler should be configured and started as part of the server.  

**USE_GITHUB**  
_Default: 0_  
A boolean value to communicate whether or not the GitHub handler should be configured and started as part of the server.  


CHANGELOG.md
# Changelog For ProboCI Open Source Server

Please note that this file can also be seen on the WIKI page for this project at:  
https://github.com/ElusiveMind/probo-docker/wiki/Changelog-For-ProboCI-Open-Source-Server

#### Last Updated: June 24, 2019

v.22 - June 24, 2019
  - Resolved a few conflicts in code caused by merge policy.

v.21 - June 23, 2019
  - Added Perl handler for installing CPAN modules
  - Multiple versions of PHP handled in a single PHP container

v.20 - June 17, 2019
  - Added modified Asset Receiver that allows for custom endpoints (MinIO Support)

v.19 - June 15, 2019
  - Added WebhookSecret to GitLab
  - Changed ports for git handlers:
    - 3012: Bitbucket
    - 3013: GitLab
    - 3014: GitHub
  - Fixed path issue that caused container manager data to not be persistent
  - Updated images to use latest brand of mbagnall enabled ProboCI Docker images.
  - Fixed mis-named variables in README (documentation)
  - Added API token to configurations of get repositories. 

v.18 - November 11, 2018
  - Added the BYPASS_TIMEOUT variable to prevent builds from being torn down on a timeout.
  - Added support for custom mbagnall images.
    - MailCatcher support
    - PostgreSQL support
    - Memcached support
  - Fix issues with loom configuration files and logging.
  - Added a flag which sill suppress loom logs on configuration.
  - Fix SOLR startup commands.

v.17 - June 13, 2018
  - Updated to pull in the latest updates from the probo project for the Drupal plugin.

v.16 - March 11, 2018
  - Corrected inconsistent environment variable naming in yml files.

v.14 - January 14, 2018
  - Optimization of Dockerfile to reduce the overall size of the image. Remove a bunch of redundant commands.
  - Updated example docker-compose.yml file with proper boot order of images. Hopefully stabilizes startup process.
  - Removed default cron job because it did not work. Needs to be set up from host vm now. Will provide documentation on wiki.
  - Finalizing v.14 as the first build candidate for general use as part of the Drupal 8 module. 

v.13 - January 13, 2018
  - Implementation of a default cron job to run probo-reaper every 5 minutes.
  - Begin documentation of environment variables in README for display on DockerHub.
  - Removed several environment configurations that do not need to be configurable for the sake of simplicity.

v.12 - January 9, 2018
  - Convert logging path to environment variable which can be set in docker call or docker-compose.
  - Clean up of environment variable names to be in line with daemon names.
  - Allow for configuration of logging mode (on/off)

v.11 - January 8, 2018
  - Modified forked probo to provide only data we need to Drupal. Data pushed to Rest API (Drupal) if configured.

v.10 - January 7, 2018
  - Optimization of container by removing unnecessary packages from being installed
  - Change to the forked version of probo which provides status updates to Drupal's dashboard module.

v.09 - January 6, 2018
  - Removed default.yaml file so that default tokens would not be included in configuration.
  - Fixed remaining formatting bug with Loom tokens. Tokens now confirmed working with Loom.
  - Allow the proxy port to be configurable.

v.08 - January 1, 2018
  - Worked on token system with loom. No token configuration if no token set in configuration.
  - Worked on token system with asset receiver. No token configuration if no token set in configuration.
  - Some bash scripting cleanup (and debugging)

v.07 - December 28, 2017
  - Implemented a public Loom service with a token. Token strongly suggested since this can noe be accessed publically.
  - Set permissions on log file paths to proper owners and read/write permissions.

v.06 - December 27, 2017  
  - Add GitLab handler code. (Currently patched due to case-sensitivity bug).
  - Moved Container Manager and Bitbucket handler yaml code into template replacement like the the other yaml templates.
  - Made the inclusion of git handlers configurable based on environment variables. They are  only included if they are specified in the docker-compose.yml file.

v.05 - December 25, 2017
  - Added checks for various database, file and data file storage directory environmental vars for the purposes of data persistence into volumes.
  - Added instructions into the README regarding the importance of persistence of database files  and other files especially asset files not hosted on S3.
  - Added a CHANGELOG file that keeps lists of changes instead of keeping all of these in the README. README now limited to the last three updates. CHANGELOG contains everything.

v.04 - December 21, 2017
  - Remove web components from Probo server container. Separate web portal to its own container.
  - Change all messagging to Kafka in preparation for Probo Notifier
  - Disabled web admin port (8080) for RethinkDB so we can repurpose.
  - Added Probo Notifier service and configured for Kafka (untested)
  - Enabled Probo Reaper service and configured for Kafka (untested)
  
v.03 - December 20, 2017
  - Modified probo-proxy so it will properly point to the correct place for the container on the correct host
  - Allowed for configuration of the correct host as per above in the probo-proxy yml file.

v.02 - December 19, 2017
  - Fixed several configuration naming and path issues.
  - Search and replace environment names in configuration files with their values
  - Fly in patched version of probo-request-logger until new version incorporates fixes friendly to open source.
  - Fixes to exposed ports to matched our exported and public services

v.01 - December 18, 2017
  - Successfully built Container Manager.
  - Successfully built Bitbucket Handler (without complete configuration).
  - Successfully built Asset Manager.
  - Successfully built Proxy (run as root).
  - Successfully built Loom (with RethinkDB as a back end).
  - Successfully built Reaper although currently unimplemented.
  - Successfully built and configured RethinkDB.

sh/startup.sh
#!/bin/sh

# this is the script which processes all of our configuration files and
# assigns our configured environmental variables to their correct locations
# in those configuration files.

# Make it so the docker socket can be read.
chmod 777 /var/run/docker.sock

if [[ -n "${LOOM_SERVER_TOKEN}" ]]; then
  export PROBO_LOOM_SERVER_TOKEN="${LOOM_SERVER_TOKEN}"
  export LOOM_SERVER_TOKEN="tokens:"$'\n'"  - ${LOOM_SERVER_TOKEN}"
fi

if [[ ! -z "${CREATE_LOOM_TASK_LOGS}" ]] && [ $CREATE_LOOM_TASK_LOGS = "1" ]; then
  export LOOM_STORAGE_DATA_DIR="storageDataDir: \"/opt/probo/data/database/loom\""
else
  export LOOM_STORAGE_DATA_DIR=""
fi

if [[ -n "${ASSET_RECEIVER_TOKEN}" ]]; then
  export PROBO_ASSET_RECEIVER_TOKEN="${ASSET_RECEIVER_TOKEN}"
  export ASSET_RECEIVER_TOKEN="tokens: '${ASSET_RECEIVER_TOKEN}'"
fi

# Create our data directories and set permissions
mkdir -p -m 777 /opt/probo/data/database/loom
chown probo:probo /opt/probo/data/database/loom

mkdir -p -m 777 /opt/probo/data/container-manager-data
chown probo:probo /opt/probo/data/container-manager-data

mkdir -p -m 777 /opt/probo/data/databases/asset-receiver
chown probo:probo /opt/probo/data/databases/asset-receiver

mkdir -p -m 777 /opt/probo/data/files/asset-receiver
chown probo:probo /opt/probo/data/files/asset-receiver

mkdir -p -m 777 /opt/probo/data/database/rethinkdb
chown probo:probo /opt/probo/data/database/rethinkdb

# Substitute environment variables from docker-compose.yml into our yml files.
# TODO: Make sure all required variables have a valid value.
envsubst < /opt/probo/yml/assets-default.yml > /opt/probo/probo-asset-receiver/defaults.config.yaml
envsubst < /opt/probo/yml/probo-defaults.yml > /opt/probo/probo/defaults.yaml
envsubst < /opt/probo/yml/container-manager.yml > /opt/probo/probo/container-manager.yml
envsubst < /opt/probo/yml/loom-defaults.yml > /opt/probo/probo-loom/loom.yml
envsubst < /opt/probo/yml/proxy-defaults.yml > /opt/probo/probo-proxy/proxy.yml
envsubst < /opt/probo/yml/reaper-defaults.yml > /opt/probo/probo-reaper/reaper.yml
#envsubst < /opt/probo/yml/notifier-defaults.yml > /opt/probo/probo-notifier/defaults.yaml
#cp /opt/probo/yml/loom-default-defaults.yml /opt/probo/probo-loom/defaults.yaml

# start rethinkdb before we start the loom service and run it as probo to
# avoid permissions problems.
cd /opt/probo/data/database/rethinkdb
rethinkdb --daemon --no-http-admin --runuser probo --rungroup probo

# logging directory for process logging if configured to do so. create the directory
# anyway. it will just be empty if not configured for use.
mkdir -p -m 777 /opt/probo/data/logs
chown probo:probo /opt/probo/data/logs

# start all of our probo processes as the probo user with the exception of
# the proxy in case the proxy is run on port 80.
if [[ ! -z "${PROBO_LOGGING}" ]] && [ $PROBO_LOGGING = "1" ]; then
  su - probo -c "/opt/probo/probo/bin/probo container-manager -c /opt/probo/probo/container-manager.yml > /opt/probo/data/logs/container-manager.log &"
else
  su - probo -c "/opt/probo/probo/bin/probo container-manager -c /opt/probo/probo/container-manager.yml &"
fi

if [[ ! -z "${USE_GITHUB}" ]] && [ $USE_GITHUB = "1" ]; then
  envsubst < /opt/probo/yml/github-defaults.yml > /opt/probo/probo/github-handler.yml
  if [[ ! -z "${PROBO_LOGGING}" ]] && [ $PROBO_LOGGING = "1" ]; then
    su - probo -c "/opt/probo/probo/bin/probo github-handler -c /opt/probo/probo/github-handler.yml > /opt/probo/data/logs/github-handler.log &"
  else
    su - probo -c "/opt/probo/probo/bin/probo github-handler -c /opt/probo/probo/github-handler.yml &"
  fi
fi

if [[ ! -z "${USE_BITBUCKET}" ]] && [ $USE_BITBUCKET = "1" ]; then
  envsubst < /opt/probo/yml/bitbucket-defaults.yml > /opt/probo/probo-bitbucket/bitbucket-handler.yml
  if [[ ! -z "${PROBO_LOGGING}" ]] && [ $PROBO_LOGGING = "1" ]; then
    su - probo -c "/opt/probo/probo-bitbucket/bin/probo-bitbucket-handler -c /opt/probo/probo-bitbucket/bitbucket-handler.yml > /opt/probo/data/logs/bitbucket-handler.log &"
  else
    su - probo -c "/opt/probo/probo-bitbucket/bin/probo-bitbucket-handler -c /opt/probo/probo-bitbucket/bitbucket-handler.yml &"
  fi
fi

if [[ ! -z "${USE_GITLAB}" ]] && [ $USE_GITLAB = "1" ]; then
  envsubst < /opt/probo/yml/gitlab-defaults.yml > /opt/probo/probo-gitlab/gitlab-handler.yml
  if [[ ! -z "${PROBO_LOGGING}" ]] && [ $PROBO_LOGGING = "1" ]; then
    su - probo -c "/opt/probo/probo-gitlab/bin/probo-gitlab-handler -c /opt/probo/probo-gitlab/gitlab-handler.yml > /opt/probo/data/logs/gitlab-handler.log &"
  else
    su - probo -c "/opt/probo/probo-gitlab/bin/probo-gitlab-handler -c /opt/probo/probo-gitlab/gitlab-handler.yml &"
  fi
fi

if [[ ! -z "${PROBO_LOGGING}" ]] && [ $PROBO_LOGGING = "1" ]; then
  su - probo -c "/opt/probo/probo-asset-receiver/bin/probo-asset-receiver > /opt/probo/data/logs/asset-receiver.log &"
  su - probo -c "/opt/probo/probo-loom/bin/loom -c /opt/probo/probo-loom/loom.yml > /opt/probo/data/logs/probo-loom.log &"
  # start the proxy as the root user just in case we're on port 80.
  node /opt/probo/probo-proxy/index.js -c /opt/probo/probo-proxy/proxy.yml > /opt/probo/data/logs/probo-proxy.log &
else
  su - probo -c "/opt/probo/probo-asset-receiver/bin/probo-asset-receiver &"
  su - probo -c "/opt/probo/probo-loom/bin/loom -c /opt/probo/probo-loom/loom.yml &"
  # start the proxy as the root user just in case we're on port 80.
  node /opt/probo/probo-proxy/index.js -c /opt/probo/probo-proxy/proxy.yml &
fi

su - probo -c '/opt/probo/probo-reaper/bin/probo-reaper server'


sh/node-startup.sh
#!/bin/sh

# this is the script which processes all of our configuration files and
# assigns our configured environmental variables to their correct locations
# in those configuration files.

#########################################################
# CONTAINER MANAGER CONFIGURATION FILE PROCESSOR
#########################################################
exec 3<> /opt/probo/probo/container-manager.yml
echo "hostname: localhost" >&3
echo "port: 3020" >&3
echo "" >&3
echo "instanceName: '$CM_INSTANCE_NAME'" >&3
echo "api:" >&3
echo "  url: 'http://localhost:3012'" >&3
echo "" >&3
echo "auth:" >&3
echo "  url: 'http://localhost:3012/auth_lookup'" >&3
echo "" >&3
echo "loom:" >&3
echo "  url: 'http://localhost:3060'" >&3
echo "" >&3
echo "assets:" >&3
echo "  url: '$ASSET_SERVER_URL'" >&3
echo "" >&3
echo "buildUrl: '$PROBO_BUILD_URL'" >&3
echo "endpointUrl: '$DRUPAL_ENDPOINT_URL'" >&3
su - probo -c 'exec 3>&-'

#########################################################
# BITBUCKET CONFIGURATION FILE PROCESSOR
#########################################################
exec 4<> /opt/probo/probo-bitbucket/bitbucket-handler.yml
echo "port: 3012" >&4
echo "hostname: 0.0.0.0" >&4
echo "" >&4
echo "bbWebhookUrl: '$BB_WEBHOOK_URL'" >&4
echo "" >&4
echo "# Bitbucket client key and client secret. These are generated for you" >&4
echo "# when you create a new OAuth client in Bitbucket" >&4
echo "bbClientKey: $BB_CLIENT_KEY" >&4
echo "bbClientSecret: $BB_CLIENT_SECRET" >&4
echo "bbAccessToken: $BB_ACCESS_TOKEN" >&4
echo "bbRefreshToken: $BB_REFRESH_TOKEN" >&4
echo "" >&4
echo "# Settings for the API or Container Manager server port" >&4
echo "api:" >&4
echo "  url: 'http://localhost:3020'" >&4
su - probo -c 'exec 4>&-'

# Make it so the docker socket can be read.
chmod 777 /var/run/docker.sock

envsubst < /opt/probo/yml/assets-default.yml > /opt/probo/probo-asset-receiver/asset-receiver.yml
envsubst < /opt/probo/yml/probo-defaults.yml > /opt/probo/probo/defaults.yaml
envsubst < /opt/probo/yml/ghh-defaults.yml > /opt/probo/probo/ghh.yml
envsubst < /opt/probo/yml/loom-defaults.yml > /opt/probo/probo-loom/loom.yml
envsubst < /opt/probo/yml/proxy-defaults.yml > /opt/probo/probo-proxy/proxy.yml
envsubst < /opt/probo/yml/reaper-defaults.yml > /opt/probo/probo-reaper/reaper.yml

# start rethinkdb before we start the loom service and run it as probo to
# avoid permissions problems.
cd /opt/probo/probo-loom
rethinkdb --daemon --runuser probo --rungroup probo

# start all of our probo processes as the probo user with the exception of
# the proxy in case the proxy is run on port 80.
su - probo -c '/opt/probo/probo/bin/probo container-manager -c /opt/probo/probo/container-manager.yml > /dev/null &'
su - probo -c '/opt/probo/probo-bitbucket/bin/probo-bitbucket-handler -c /opt/probo/probo-bitbucket/bitbucket-handler.yml > /dev/null &'
su - probo -c '/opt/probo/probo-asset-receiver/bin/probo-asset-receiver -c /opt/probo/probo-asset-receiver/asset-receiver.yml > /dev/null &'
su - probo -c 'mkdir /opt/probo/probo-loom/data'
su - probo -c 'chmod 777 /opt/probo/probo-loom/data'
su - probo -c '/opt/probo/probo-loom/bin/loom -c /opt/probo/probo-loom/loom.yml > /dev/null &'

# start the proxy as the root user
node /opt/probo/probo-proxy/index.js -c /opt/probo/probo-proxy/proxy.yml > /dev/null &

# start the web server
rm -rf /run/httpd/* /tmp/httpd*
exec /usr/sbin/apachectl -DFOREGROUND